Extend stop compaction api

bhalevy opened the pull request:
Allow stopping compaction by type on a given keyspace and list of tables.

Also add api unit test suite that tests the existing `stop_compaction` api
and the new `stop_keyspace_compaction` api.

Fixes #9700

nyh commented on api/compaction_manager.cc:
```diff
@@ -117,6 +121,23 @@ void set_compaction_manager(http_context& ctx, routes& r) {
+        co_await ctx.db.invoke_on_all([&ks_name, &table_names, type] (database& db) {
+            auto& cm = db.get_compaction_manager();
+            return parallel_for_each(table_names, [&db, &cm, &ks_name, type] (sstring& table_name) {
+                auto& t = db.find_column_family(ks_name, table_name);
```
I'm not familiar enough with this area of the code, so I have to ask: if find_column_family() or validate_keyspace() above throw - will this provide the user with a decent error message or a mess?

nyh replied:
By the way, thanks to this review I found a bug in Alternator, which also used this find_column_family in its UpdateTable request, so results in a funky exception when given a non-existent table instead of the appropriate error message :-)

nyh replied:
If you're curious - https://github.com/scylladb/scylla/issues/9747

bhalevy replied:
> I'm not familiar enough with this area of the code, so I have to ask: if find_column_family() or validate_keyspace() above throw - will this provide the user with a decent error message or a mess?

```
$ ./scylla-cli.py compaction_manager stop_keyspace_compaction/{keyspace} POST --keyspace foo --type CLEANUP
{'message': 'Keyspace foo Does not exist', 'code': 400}

$ ./scylla-cli.py compaction_manager stop_keyspace_compaction/{keyspace} POST --keyspace system_schema --tables foo,bar --type CLEANUP
{'message': "std::_Nested_exception<no_such_column_family> (Can't find a column family foo in keyspace system_schema): std::out_of_range ()", 'code': 500}

Not a total mess but the table-not-found error is thrown as a nested exception that isn't very elegant, like what you encountered in #9747

nyh replied:
Exactly what I worried about. Note that's not just the textual error message - it's also a different error code (400 vs 500). In Alternator the different error code is important, because 400 is a real error in the request, while 500 (internal server error) means the user should retry the request - and the DynamoDB driver indeed retries the same request multiple times.
I don't know how much of a difference it makes in our REST API because it isn't documented anywhere :-(

nyh commented on test/api/run:
```diff
@@ -0,0 +1,142 @@
+#!/bin/bash
```
I can't find any obvious error in this code, but I want to point out a second time in two days that this is yet another copy of the same code, which I am still suggesting to you to not do, and instead use the python library I wrote to do the same thing - see example in test/alternator/run. This is exactly why I wrote that Python library that both test/alternator/run and test/cql-pytest now run - so we won't have all that code duplication, and every time we need to fix a bug or add a boot-time parameter, we won't need to fix it in 7 places.

bhalevy replied:
Sorry, I followed you references to test/redis/run in https://github.com/scylladb/scylla/pull/9731
I'll convert it to Python and use the requests Python module instead of curl.

nyh replied:
Thanks. I'll also do what I preach and convert redis/run to Python today (https://github.com/scylladb/scylla/issues/9748).
To be honest, this is not my preaching - I actually liked those self-contained shell scripts when there were only one or two copies, until @kostja and others rightly asked to convert it to Python (although that is just a matter of preference) and move the main scylla-starting logic to one place instead of being duplicated in multiple scripts.

nyh replied:
I sent a patch titled "test/redis: replace run script with shorter Python script"
You can do exactly the same, I think it's very easy, or even easier (you don't even need the new "redis_cmd" because you don't even need to change Scylla's parameters - the REST API is on by default). The code is certainly much shorter than the shell script with the full logic, Scylla command line, etc.

nyh commented on test/api/run:
```diff
@@ -135,6 +144,12 @@ else
 fi
 
+basedir="$(dirname $0)"
+for test in "$basedir"/*/run.sh; do
```
I'm curious why you wanted to call those files basedir/*/run.sh instead of just basedir/*.sh or basedir/*.somethingelse. What do you expect those subdirectories will have besides the run.sh?

nyh replied:
More importantly, I would seriously consider using pytest as test/{alternator,cql-pytest,redis} do, it's so much more convenient than these test scripts, and you can see the different failures on Jenkins, and so on.

bhalevy replied:
np, will use pytest.
For basic tests I expect only run.sh, but we may want to add other files there if needed.

nyh commented on test/api/run:
```diff
@@ -99,6 +99,15 @@ query_api() {
     echo "$out"
 }
 
+# Test that scyla is serving the api port.
```
This comment is wrong (you copied it from somewhere?)

bhalevy replied:
It was adapted from the redis test, and should have been moved indeed down to the loop that's probing the api.

nyh commented on test/api/run:
```diff
@@ -99,6 +99,15 @@ query_api() {
+# Test that scyla is serving the api port.
+check_api() {
+    out=$(query_api "$@")
+    if echo "$out" | grep -q '"code"'; then
```
That's a strange idiom for matching in shell (why not use a "case", without echo or grep?) 

By the way, am I right that this is just a very basic check for the API. It just checks that the API doesn't result in an error, but nothing about what it does. Right?

bhalevy replied:
> That's a strange idiom for matching in shell (why not use a "case", without echo or grep?)

yeah, case would be good here.

> By the way, am I right that this is just a very basic check for the API. It just checks that the API doesn't result in an error, but nothing about what it does. Right?

correct

bhalevy commented on api/storage_service.cc:
```diff
@@ -69,11 +69,12 @@ const locator::token_metadata& http_context::get_token_metadata() {
-    if (ctx.db.local().has_keyspace(param["keyspace"])) {
-        return param["keyspace"];
+sstring validate_keyspace(http_context& ctx, const parameters& param, sstring param_name) {
+    auto ks_name = param[param_name];
```
sure

bhalevy commented on api/api-doc/compaction_manager.json:
```diff
@@ -112,6 +112,46 @@
+         "operations":[
+            {
+               "method":"POST",
+               "summary":"Stop all running compaction-like tasks in the given keyspace having the provided type",
```
ok, I'll tables into the description

bhalevy commented:
In v2:
- re-implemented the api tests using the cql-pytest infrastructure
  - and pulled the basic tests upfront so they can be used for regression the testing throughout the series
  - use requests python module for accessing the api
- use `const sstring&` in `validate_keyspace`
- improved `stop_keyspace_compaction` json api description
- added some negative tests to test non-exiting keyspace and table

nyh commented on test/api/conftest.py:
```diff
@@ -0,0 +1,104 @@
+from cassandra.cluster import Cluster, ConsistencyLevel, ExecutionProfile, EXEC_PROFILE_DEFAULT
+from cassandra.policies import RoundRobinPolicy
+
+# By default, tests run against the CQL server listening
```
I found myself unsure whether all the CQL-related stuff - the configuration options, the fixture, etc., is here by mistake (you just copied it) or by design - maybe you plan to add a test in the future which combines CQL requests and API requests?

By the way, if your intention is that **most** tests will combine CQL and API requests, this raises another sacrilegious question: Maybe there shouldn't be a separate REST API test suite at all? Can't we put the REST API tests in the existing cql-pytest? Note that cql-pytest already uses the REST API for some things, such as to implement "nodetool flush",  and other nodetool commands needed for some tests. Nothing really prevents us from adding pure-REST-API tests in cql-pytest.

Although maybe there's also benefit of a separate test suite, to make it more obvious that cql-pytest tests are supposed - mostly - to also work on Cassandra - while none of the REST API tests should. I don't know.

nyh replied:
Ok, I think both options are reasonable. The separate subdirectory isn't that bad now that the run script is tiny, so maybe it does make sense to separate it. The more I think about it, the more sense it makes to me.

But if you do decide to move it into cql-pytest, you can either put all of them in the same file (e.g., test_rest_api.py), or, you can also create a subdirectory - that works well too - you can see an example in test/cql-pytest/cassandra_tests and its tree of subdirectories.
In that case, check out existing code in test/cql-pytest/nodetool.py that deals with finding the REST API.  Also, regarding "scylla_only" - your new tests can use a fixture or function that returns the rest_api, and if called when the REST API is not available (see example checking code in cql-pytest/nodetool.py) it will call pytest.skip() to skip the test. So you won't need to add a "scylla_only" tag explicitly to each test.

bhalevy replied:
okay, it's less work to keep it separated rather than fighting with the cql-pytest ramifications, so I'll leave it that way :)

nyh commented on test/api/conftest.py:
```diff
@@ -0,0 +1,104 @@
+    return cluster.connect()
+
+class ApiSession:
+    def __init__(self, host='locahost', port=10000):
```
Nitpick: if you never use the default parameter values, I think it's better not to have them at all. It's just confusing for readers who may think these defaults are used for something.

bhalevy replied:
okay (and there's a typo hidden there...)

nyh commented on test/api/conftest.py:
```diff
@@ -0,0 +1,104 @@
+    def __init__(self, host='locahost', port=10000):
+        self.host = host
+        self.port = port
+        self.session = Session()
```
nitpick: I'd use "import requests" and use `requests.Session()` here explicitly. I found it very non-obvious where this "Session" comes from, it's not a very self-explanatory name.

bhalevy replied:
can do

nyh commented on test/api/run:
```diff
@@ -0,0 +1,31 @@
+pid = run.run_with_temporary_dir(cmd)
+ip = run.pid_to_ip(pid)
+
+run.wait_for_cql(pid, ip)
```
Although this will work in practice - because we (?) start the REST API before CQL (do we?) - the better approach would be to also check the REST API came up. You can see an example of how to check both CQL and something else in the new patch I sent for redis. Or in the already existing test/alternator/run.

bhalevy replied:
We do start listening on the API before CQL (one reason is to be able to stop reshape compaction when populating tables in the distributed_loader), but it will make no harm if we check for both CQL and REST API...

nyh commented on test/api/test_system.py:
```diff
@@ -0,0 +1,20 @@
+# You should have received a copy of the GNU Affero General Public License
+# along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+
+def test_system_uptime_ms(cql, api):
```
If this specific test doesn't need "cql", you don't need it in the argument list.
I still don't know if you plan any tests that will need CQL, but this one certainly doesn't.

bhalevy replied:
additional tests added later in the series do use it.

nyh replied:
Ok, I understand that now, but in this case, this specific test "test_system_uptime_ms" doesn't need the cql fixture, so it shouldn't list it.

nyh commented:
BTW I'd also like to get a review from @raphaelsc on whether the actual feature is useful (beyond all the implementation details I reviewed). The issue (#9700) doesn't really explain why this is a useful feature or who would use it.

bhalevy commented:
In v3:
- Added:
  - database: un-nest no_such_keyspace and no_such_column_family exceptions
  - api: storage_service: validate_keyspace: improve error message
  - api: storage_service: add validate_tables
- test/rest_api
  - renamed api to rest_api in new test
  - improved documentation
  - cleaned up fixture dependencies
  - added `check_rest_api`

nyh commented on test/cql-pytest/run.py:
```diff
@@ -260,6 +261,15 @@ def check_ssl_cql(ip):
     ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
     check_cql(ip, ssl_context)
 
+## Test that the Scylla REST API is serving, for wait_for_services() below.
```
By the way, I was tempted to ask you to move this function to test/rest-api/run where you used it, but then I changed my mind - we should perhaps even use it in test/cql-pytest/run as well so make sure that the REST API is also up. Although maybe it's just a waste of time? I don't know.

bhalevy replied:
I think it's a good idea to use it in cql-pytest since it does use the api.
This will verify that the rest api is up and running along with the cql service.
It's possible (although highly unlikely) that a future change will start CQL before the REST API.

bhalevy replied:
Done in #9755

nyh commented on test/rest_api/README.md:
```diff
@@ -0,0 +1,35 @@
+Tests use the requests library and the pytest frameworks
+(both are available from Linux distributions, or with "pip install").
+
+To run all tests using test.py, just run `./test.py api/run`.
```
is it now rest_api/run?

Also I would just run "test/rest_api/run". No need to ask test.py to do it ;-) Although you're right, you can.

nyh replied:
Also this line and the "more conveniently, ..." paragraph below talk about basically the same thing.

bhalevy replied:
I never remember the pytest details as I rarely use it, hence the ./test.py instructions for dummies like me.

bhalevy replied:
api -> rest_api fixed in the next version

bhalevy replied:
also in https://github.com/scylladb/scylla/issues/9754 that will introduce it

bhalevy replied:
Done in #9755

nyh commented on test/rest_api/conftest.py:
```diff
@@ -24,10 +24,20 @@
+from cassandra.policies import RoundRobinPolicy
+
+sys.path.insert(1, sys.path[0] + '/../cql-pytest')
+from util import unique_name
```
It's weird to duplicate so much code from cql-pytest/conftest.py and then use a 5-line function from util.py ;-) I wonder if we could import the full conftest.py from cql-pytest. Or maybe not all of it is relevant?
But we don't have to do this now.

bhalevy replied:
that's an interesting approach.
maybe we should have a formal scylla-pytest python module that will be easy to consume by the different tests (alternator, cql-pytest, redis, and now rest_api)

nyh commented on test/rest_api/test_compaction_manager.py:
```diff
@@ -15,6 +15,50 @@
+    try:
+        resp.raise_for_status()
+    except HTTPError:
+        pass
```
nitpick:
Instead of all this raise_for_status() thing, the more idiomatic pytest (and shorter code) would be to do something like
```python
assert resp.status_code = 400
```
or 
```python
assert !resp.ok
```
or something.

bhalevy replied:
there are named status codes that we can check although in some cases the test doesn't care about the exact status_code (currently we do) so checking for the error "class" via the exception type would be more useful.

bhalevy replied:
Done in #9755

nyh commented on test/rest_api/test_compaction_manager.py:
```diff
@@ -15,6 +15,50 @@
+
+def test_compaction_manager_stop_keyspace_compaction_tables(cql, rest_api, test_keyspace):
+    with new_test_table(cql, test_keyspace, "a int, PRIMARY KEY (a)") as t0:
+        test_tables = [t0 if not '.' in t0 else t0.split('.')[1]]
```
why this if? t0 will always, as returned by new_test_table, contain the keyspace dot table name.

bhalevy replied:
until it breaks :)

bhalevy replied:
done

nyh replied:
What was done in 9755? Did you put this comment in the wrong thread?
But never mind, the unnecessary "if" will not hurt anyone. It's ok.

nyh commented on test/rest_api/test_compaction_manager.py:
```diff
@@ -15,6 +15,50 @@
+
+        # multiple tables
+        with new_test_table(cql, test_keyspace, "b int, PRIMARY KEY (b)") as t1:
+            test_tables += [t1 if not '.' in t1 else t1.split('.')[1]]
```
ditto - unnecessary if, will always be the else case

bhalevy replied:
done

nyh reviewed (changes requested):
Looks good, but I left a few more comments and requests (sorry ;-)).

bhalevy commented:
Depends on https://github.com/scylladb/scylla/pull/9755

bhalevy commented on database.cc:
```diff
@@ -966,8 +966,8 @@ future<> database::drop_column_family(const sstring& ks_name, const sstring& cf_
-    } catch (...) {
-        throw std::out_of_range("");
+    } catch (std::out_of_range&) {
+        throw no_such_column_family(ks, cf);
```
Good catch. Will do.

bhalevy replied:
> Please change the comment in database.hh that claims (now wrongly) that this function "throws std::out_of_range if missing"

fixed in b28314c2e51bda465670faf59a9c211f47d10b13

bhalevy replied:
> Also, while I really like this patch, I think we can't do it without looking for existing code which might rely on the specific out_of_range that this old code used to return.

Done in #9755

bhalevy commented on database.cc:
```diff
@@ -1024,34 +1024,26 @@ std::vector<lw_shared_ptr<column_family>> database::get_non_system_column_famili
-    } catch (...) {
-        std::throw_with_nested(no_such_column_family(ks_name, cf_name));
-    }
+    return find_column_family(find_uuid(ks_name, cf_name));
```
As I explained in https://github.com/scylladb/scylla/issues/9754#issue-1073511639
I think it's fine to return the bad param error initially when validating the request parameters and then, if the table goes away, or for any other reason we fail to locate it - it's indeed an internal server error (that doesn't care to lock anything while performing the api call)

bhalevy replied:
since these calls are synchronous, it should be an internal error to not find the uuid returned by `find_uuid(ks_name, cf_name)`.  However, currently find_column_family(uuid) throws no_such_column_family(uuid) which is not the best error since it hides the original ks_name, cf_name; but I think we can live with it.

bhalevy replied:
Done in #9755

bhalevy commented on database.cc:
```diff
@@ -1268,11 +1260,7 @@ std::vector<view_ptr> keyspace_metadata::views() const {
-    } catch (std::out_of_range&) {
-        std::throw_with_nested(no_such_column_family(ks_name, cf_name));
-    }
+    return find_schema(find_uuid(ks_name, cf_name));
```
`find_schema(uuid)` calls `find_column_family(uuid).schema()` so it will throw `no_such_column_family(uuid)` in this case.  Ditto as above, not ideal, but would work.

bhalevy replied:
Done in #9755

bhalevy commented on test/cql-pytest/run.py:
```diff
@@ -260,6 +261,15 @@ def check_ssl_cql(ip):
 
+## Test that the Scylla REST API is serving, for wait_for_services() below.
+def check_rest_api(ip, port=10000):
+    resp = requests.get(f"http://{ip}:{port}/ui")
```
What is the top level url?
`http://{ip}:{port}/` return Error 404.

I wanted to use a neutral url that isn't of any actual api so it's unlikely to change.

bhalevy replied:
I don't care that it returns `301 Moved Permanently` as it's just used to probe the api port.
The error proves that the server is up and running.

bhalevy replied:
Done in #9755

bhalevy commented on test/cql-pytest/run.py:
```diff
@@ -260,6 +261,15 @@ def check_ssl_cql(ip):
+## Test that the Scylla REST API is serving, for wait_for_services() below.
+def check_rest_api(ip, port=10000):
+    resp = requests.get(f"http://{ip}:{port}/ui")
+    try:
```
nope, requests.get will succeed and return an erroneous response.
resp.raise_for_status() raises the exception, not requests.get().

bhalevy replied:
Done in #9755

bhalevy commented on test/cql-pytest/run.py:
```diff
@@ -260,6 +261,15 @@ def check_ssl_cql(ip):
+    resp = requests.get(f"http://{ip}:{port}/ui")
+    try:
+        resp.raise_for_status()
+    except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError):
```
can do

bhalevy replied:
Done in #9755

bhalevy commented on api/storage_service.cc:
```diff
@@ -77,6 +77,18 @@ sstring validate_keyspace(http_context& ctx, const parameters& param, sstring pa
     throw bad_param_exception(fmt::format("Keyspace '{}' does not exist", ks_name));
 }
 
+std::vector<sstring> validate_tables(const sstring& ks_name, http_context& ctx, const parameters& param, sstring param_name) {
```
> So again we go back to the wrong exception...

I don't think it's wrong to return internal_server_error code if the tables exist when the api function is called and later disappears.  It's certainly can't be considered a bad request.  If it's retried on the client, the retry invocation will get the bad_request error since now the tables are gone.

> So isn't it better to just add the validation inside the loop that works on each individual table, and not as an advanced step going over all the tables?

In the case that the user provided no tables option we get the list of tables from the database, and they can essentially go away too if this is done with a DROP TABLE in parallel.

I think it's going to be super weird for a user to run an operation on the whole keyspace and get a bad_request error for an internal error like this.

The right thing to do is to lock the keyspace in such a way that will prevent tables from being dropped in the middle of the operation - OR provide a way to abort the operation internally if a table is dropped during an api operation, log that, and return an appropriate error to the user.

I think that for short-running operations like enable/disable auto_compaction, locking is fine.
For long-running operations, like running cleanup compaction, we need first to convert the api to asynchronous (generalization of https://github.com/scylladb/scylla/issues/8392) and then the operation can be managed more carefully and be aborted (table removal does stop ongoing compactions, including cleanup, I think this need not generate an error if there's a background cleanup operation on it)

bhalevy replied:
Done in #9755

bhalevy commented:
In v4:
- Rebased on top of e032f92c5cf400c6bdc253c0dbba735ce4aafe84 (#9755)
  - uses `parse_tables`

bhalevy commented on database.cc:
```diff
@@ -1024,34 +1024,26 @@ std::vector<lw_shared_ptr<column_family>> database::get_non_system_column_famili
-    } catch (...) {
-        std::throw_with_nested(no_such_column_family(ks_name, cf_name));
-    }
+    return find_column_family(find_uuid(ks_name, cf_name));
```
Done in #9755

bhalevy commented:
@nyh, can you please re-review now that #9755 was merged?

nyh commented on test/rest_api/test_compaction_manager.py:
```diff
@@ -15,6 +15,69 @@
+def new_keyspace(cql, this_dc):
+    name = unique_name()
+    cql.execute(f"CREATE KEYSPACE {name} WITH REPLICATION = {{ 'class' : 'NetworkTopologyStrategy', '{this_dc}' : 1 }}")
+    return name
```
The more "proper" way to do this would be to use a "contextmanager" that willl create a new keyspace and return its name - but make sure the keyspace is deleted when the test ends - regardless of whether the test failed or succeeded. You can see an example of `new_test_table()` in test/cql-pytest/util.py.
But I'll let this slide and approve this PR. Unlike Alternator tests where leaving an undeleted table on AWS is bad, here the result of forgetting to delete a table is less disasterous. But it's still good practice to always think about this.
